cds-snc/platform-forms-client
# [WIP] feat: Self-serve Manage Form Ownership

# Summary | Résumé

Self-serve Manage Form Ownership

| Access from Responses page  | Manage form access dialog    | Invite form |
| ----------------------------------------------- | -------------------------------------------- | ----- |
| <img width="1097" alt="image" src="https://github.com/user-attachments/assets/d683c0b6-5d46-41b9-bba1-13b45f459acf"> | <img width="728" alt="image" src="https://github.com/user-attachments/assets/7f7db1b5-06a5-4cb9-9f94-b884ad3b2359"> | <img width="711" alt="image" src="https://github.com/user-attachments/assets/e08e6942-740e-477b-ab4f-15baa89bf4e3"> |

| Re-invite | Cancel invitation |
| ------------------------- | ----------------- |
|  ![Screen Recording 2024-09-26 at 4 21 49 PM](https://github.com/user-attachments/assets/c04eb55a-bd94-4f3c-9f97-32052dbc31a4) | ![Screen Recording 2024-09-26 at 4 26 24 PM](https://github.com/user-attachments/assets/3f9977d5-5d44-4ef3-949c-e525d53628de) |

| Remove existing owner | Adding invitees + validation |
| ----- | ----- | 
| ![Screen Recording 2024-09-26 at 4 27 47 PM](https://github.com/user-attachments/assets/2b346211-1576-47f0-87b8-0898369643b2) | ![Screen Recording 2024-09-26 at 4 35 24 PM](https://github.com/user-attachments/assets/ca84cac3-e07f-4759-a738-ed829ecc16c7) |

| Accept or decline an invitation |
| ----- |
| ![Screen Recording 2024-09-26 at 4 57 28 PM](https://github.com/user-attachments/assets/533edcc7-f8ef-46d3-a2e5-577865370acc) |

## Test instructions

Using email aliases (ie, firstname.lastname+alias@cds-snc.ca) can be very helpful in testing this PR. If you can run the service locally, consider the Notify Interceptor.

Note that invitations are tied to a user's email address, and can be sent to someone whether they have an existing account or not.

When adding invitees, the following validations are run on email addresses:
- Must be from same domain as logged-in user
- Must be a government email address
- Must not already have access to the form

### Inviting a non-Forms user
- If an invitee does not have an account already, they will be sent an email referring them to the Registration form.
- After successfully registering they will land on their /forms page, where all invitations are displayed.
- If they accept the invitation, a notification goes to all existing owners to notify of the ownership change.

### Inviting a Forms user
- If an invitee already has an account, they will be sent an email instructing them to login to Forms to accept the invitation
- After successfully logging in, they will land on their /forms page, where all invitations are displayed.
- If they accept the invitation, a notification is sent to all existing owners to notify of the ownership change.

### Removing an owner
In the Manage Ownership dialog, existing owners are displayed. To remove an owner, press the Delete button and then Confirm that you want to delete the user. 

You will not be able to remove an Owner if:
- That owner is you
- There is only one owner remaining on the form

When an owner is removed, notifications are sent to the remaining owners, as well as the owner that was removed to confirm the ownership change.

### Cancelling an invitation
In the Manage Ownership dialog, previously sent invitations are displayed. To cancel an invitation, press the Cancel button and Confirm that you want to cancel. The Invitation is deleted, and no notifications are sent.

### Re-sending expired or lost invitations
In the Manage Ownership dialog, invitations that have been expired will be identified as "expired." To re-send any notification, press the resend button and the email address will be added to the Invitee list, then you can continue on the regular invitation process
